### PR TITLE
Clean dependencies

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

This PR cleans up TRL dependencies by better dividing and following best practice. 
It updates the test workflow accordingly.

- For a test extra, it's usually best practice to include only the dependencies required for testing and not all optional dependencies. This PR makes the appropriate change.
- Optional dependencies `deepspeed` and `liger-kernel` are only supported for linux. This PR ensure that they are not part of optional dependencies with windows.
- Add test for python 3.12 (stable for more than a year now)
- Test with both latest and dev versions to better track breaking changes in our dependencies.

| | OS | Python | Required dependencies | Optional dependencies |
|-|-|-|-|-|
| Tests | Ubuntu | 3.9 | Latest | Latest |
| Tests | Ubuntu | 3.10 | Latest | Latest |
| Tests | Ubuntu | 3.11 | Latest | Latest |
| Tests | Ubuntu | 3.12 | Latest | Latest |
| Tests | Windows | 3.9 | Latest | Latest |
| Tests | Windows | 3.10 | Latest | Latest |
| Tests | Windows | 3.11 | Latest | Latest |
| Tests | Windows | 3.12 | Latest | Latest |
| Tests without optional dependencies| Ubuntu | 3.12 | Latest | No |
| Tests with dev dependencies | Ubuntu | 3.12 | Dev | Latest |

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/judges.py [hunk 1]
```diff
@@ -21,7 +21,6 @@
 import numpy as np
 from accelerate import Accelerator
 from huggingface_hub import InferenceClient
-from scipy.special import softmax
```
Comment (qgallouedec) [inline trl/trainer/judges.py:24]: remove this dep

File: tests/test_peft_models.py [hunk 2]
```diff
@@ -95,7 +98,7 @@ def test_create_peft_model_from_config(self):
         nb_trainable_params = sum(p.numel() for p in trl_model.parameters() if p.requires_grad)
         assert nb_trainable_params == 10273
 
-    @require_bitsandbytes
+    @require_torch_gpu_if_bnb_not_multi_backend_enabled
```
Comment (qgallouedec) [inline tests/test_peft_models.py:101]: it requires cuda

**Comment by qgallouedec** [comment]
Failing test are expected and will be solved later (in #2288 for example)

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -15,76 +15,101 @@ on:
...
-          submodules: recursive
       - name: Set up Python ${{ matrix.python-version }}
         uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
-      - uses: pre-commit/action@v3.0.1
+          cache: "pip"
+          cache-dependency-path: |
+              setup.py
+              requirements.txt
```
Comment (lewtun) [inline .github/workflows/tests.yml:38]: If we find the dependencies are a hassle to maintain, we might want to switch to the `transformers` approach, which uses a look-up table to give fine-grained control: https://github.com/huggingface/transformers/blob/main/setup.py

File: trl/trainer/judges.py [hunk 4]
```diff
@@ -239,7 +238,11 @@ def judge(
             ranks[flip_mask] = ranks[flip_mask][:, ::-1]
 
         # Return the ranks or score probability
-        return softmax(ranks, axis=-1)[:, 0].tolist() if return_scores else ranks[:, 0].tolist()
+        if return_scores:
+            probs = np.exp(ranks) / sum(np.exp(ranks))
```
Comment (lewtun) [inline trl/trainer/judges.py:242]: I think we should use the shifted variant to avoid numerical instability: https://github.com/scipy/scipy/blob/92d2a8592782ee19a1161d0bf3fc2241ba78bb63/scipy/special/_logsumexp.py#L234

**Review by lewtun — APPROVED** [review]
LGTM with a small tweak to the softmax

File: trl/trainer/judges.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/judges.py:242]:
```python
>>> from scipy import softmax
>>> ranks = [[0.1, 2.2], [0.3, 3.4], [0.5, 1.6]]
>>> softmax(ranks, axis=-1)[:, 0]
array([0.10909682, 0.04310725, 0.24973989])
>>> exp_rank = np.exp(ranks - np.max(ranks))
>>> probs = exp_rank / exp_rank.sum(axis=0)
>>> probs[:, 0]
array([0.2693075 , 0.32893292, 0.40175958])
```

The max computation may not be correct. I'll directly take scipy source code

```suggestion
            logit_max = np.amax(ranks, axis=-1, keepdims=True)
            exp_logit_shifted = np.exp(ranks - logit_max)
            probs = exp_logit_shifted / np.sum(exp_logit_shifted, axis=-1, keepdims=True)
```

Comment (qgallouedec) [reply trl/trainer/judges.py]: done in https://github.com/huggingface/trl/pull/2298#discussion_r1824164828
